src/server/replica.cc

```diff
@@ -447,61 +413,214 @@ error_code Replica::InitiatePSync() {
 
   LOG(INFO) << "Full sync finished";
-
   state_mask_ |= R_SYNC_OK;
 
   return error_code{};
 }
 
+error_code Replica::ConsumeRedisStream() {
+  base::IoBuf io_buf(16_KB);
+  parser_.reset(new RedisParser);
+
+  ReqSerializer serializer{sock_.get()};
+
+  // Master waits for this command in order to start sending replication stream.
+  RETURN_ON_ERR(SendCommand("REPLCONF ACK 0", &serializer));
+
+  VLOG(1) << "Before reading repl-log";
+
+  // Redis sends eiher pings every "repl_ping_slave_period" time inside replicationCron().
+  // or, alternatively, write commands stream coming from propagate() function.
+  // Replica connection must send "REPLCONF ACK xxx" in order to make sure that master replication
+  // buffer gets disposed of already processed commands.
+  error_code ec;
+  time_t last_ack = time(nullptr);
+  string ack_cmd;
+
+  // basically reflection of dragonfly_connection IoLoop function.
+  while (!ec) {
+    io::MutableBytes buf = io_buf.AppendBuffer();
+    io::Result<size_t> size_res = sock_->Recv(buf);
+    if (!size_res)
+      return size_res.error();
+
+    VLOG(1) << "Read replication stream of " << *size_res << " bytes";
+    last_io_time_ = sock_->proactor()->GetMonotonicTimeNs();
+
+    io_buf.CommitWrite(*size_res);
+    repl_offs_ += *size_res;
+
+    // Send repl ack back to master.
+    if (repl_offs_ > ack_offs_ + 1024 || time(nullptr) > last_ack + 5) {
+      ack_cmd.clear();
+      absl::StrAppend(&ack_cmd, "REPLCONF ACK ", repl_offs_);
+      RETURN_ON_ERR(SendCommand(ack_cmd, &serializer));
+    }
+
+    ec = ParseAndExecute(&io_buf);
+  }
+
+  VLOG(1) << "ConsumeRedisStream finished";
+  return ec;
+}
+
+error_code Replica::ConsumeDflyStream() {
+  ReqSerializer serializer{sock_.get()};
+  // TBD
+  serializer.SendCommand("QUIT");
+  state_mask_ &= ~R_ENABLED;  // disable further - TODO: not finished.
+  RETURN_ON_ERR(serializer.ec());
+
+  base::IoBuf io_buf{128};
+
+  RETURN_ON_ERR(Recv(sock_.get(), &io_buf));
+
+  return error_code{};
+}
+
+error_code Replica::StartAsDflyFlow() {
+  CHECK(!sock_);
+  DCHECK(!master_context_.master_repl_id.empty() && !master_context_.dfly_session_id.empty());
+
+  ProactorBase* mythread = ProactorBase::me();
+  CHECK(mythread);
+
+  sock_.reset(mythread->CreateSocket());
+  RETURN_ON_ERR(sock_->Connect(master_context_.endpoint));
+
+  VLOG(1) << "Sending on flow " << master_context_.master_repl_id << " "
+          << master_context_.dfly_session_id << " " << master_context_.dfly_flow_id;
+
+  ReqSerializer serializer{sock_.get()};
+  auto cmd = StrCat("DFLY FLOW ", master_context_.master_repl_id, " ",
+                    master_context_.dfly_session_id, " ", master_context_.dfly_flow_id);
+  RETURN_ON_ERR(SendCommand(cmd, &serializer));
+
+  parser_.reset(new RedisParser{false});  // client mode
+
+  std::unique_ptr<base::IoBuf> io_buf{new base::IoBuf(128)};
+  unsigned consumed = 0;
+  RETURN_ON_ERR(ReadRespReply(io_buf.get(), &consumed));  // uses parser_
```

Comment:
I don't really like the whole we-are-throwing-bytes-around-manually vibe. A parsed response with RAII semantics (drop or re-assign) would be really handy... But its likely over-engineering

Reply:
the moment you get me a PoC with "set" only replication that includes full-sync and the journal - you do whatever refactorings you want plus 👸🏼 to marry plus half 👑 .

Reply:
The PoC is in the replication branch 😎  I'll do it sometime later. Its just the more hard code we have, the more efforts we need in the future to fix it all